proposals/0530-strict-box-denesting.rst

```diff
@@ -0,0 +1,613 @@
+Strict Box Denesting
+====================
+
+.. author:: Matthew Craven
+.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/530>`_.
+.. sectnum::
+.. contents::
+
+
+..
+  This introduction shares a lot with the proposed change specification.
+  When editing one, it is wise to consider editing the other!
+
+This proposal seeks to add a new optimization flag, tentatively named
+``-fstrict-box-denesting``, which (when enabled) allows GHC to elide
+the allocation and traversal of all heap objects created by some data
+constructors.  To see what that means, compare the following two examples:
+
+::
+
+   data    Example1 a = MkExample1 { exampleField1 :: !a }
+   newtype Example2 a = MkExample2 { exampleField2 ::  a }
+
+Today, any calls to ``MkExample1`` that remain after simplification
+will allocate a heap object at run-time, while calls to ``MkExample2``
+are no-ops.  As a result, a call to ``exampleField1`` requires
+pointer-chasing to access the field stored in that heap object, while
+a call to ``exampleField2`` is again a no-op.  But the semantic
+differences between their declarations are minor, and do not actually
+prevent a newtype-like elision of the ``MkExample1`` constructor at
+run-time.
+
+Under this proposal, when ``-fstrict-box-denesting`` is enabled, GHC
+will automatically perform this newtype-like elision for constructors
+that satisfy the following four conditions:
+
+1. Among the constructor's fields (after unpacking, counting
+   existentials and constraints), exactly one field has
+   non-zero-width.  (Otherwise, no field alone has all of the relevant
+   information.)
+2. That field's type is represented by a GC-managed pointer.
+   (Otherwise, passing it in lieu of a constructed heap object makes no
+   sense.)
+3. If that field is of lifted type, it must be a strict field.
+   (Otherwise, ``undefined`` and ``WithLazyField undefined`` would
+   need to be distinct.)
+4. Either
+
+   1. There is no other constructor in the relevant ``data`` or
+      ``data instance`` declaration, or
+   2. The underlying heap object for any evaluated expression with the
+      field's type has the same tag as the constructor.
+
+Condition 4 ensures that pattern-matching remains efficient.
+Condition 4(ii) is checked on a best-effort basis.
+
+Motivation
+----------
+
```

Comment:
There is a missing link in the narrative here:
```
The obvious question is: why not just use a `newtype`?  That is, instead of `-fstrict-box-denesting` and
   data T a = MkT !a
why not say
   newtype T a = MkT a

That is very much what newtypes are for!
```
Now you can move on to explain why newtypes don't necessarily cut it.